### ISSUE TITLE
Problem with encoding scheme?

### ISSUE THREAD

**Issue by schazers** [EnvelopSound/EnvelopForLive#15]
Hey @ramagottfried , 

Natasha Barrett and I (she says hello!) just spent a while trying to use envelop in the CCRMA listening room (24.4 system), and we may have uncovered issues in the envelop encoder. The listening room has no problem decoding ACN formatted encodings, so we know the channel orderings are fine. But it does not properly decode input from the envelop client.

CCRMA's listening room has its own ambisonic decoders running through OpenMixer, and we are using its ACN decoder. When we pass the inputs to that decoder, they are the untouched 16 channels being piped straight past the envelop decoder (16 ins -> 16 outs, untouched). 

We put together some more targeted tests, and here are what the problems seem to be:

1) The input levels from the encoder to the decoder appear to be inconsistent. Imagine panning a single source sound around the x-y plane, e.g. at points (x, y, z) = (1, 0, 0), (0, 1, 0), (-1, 0, 0), (0, -1, 0). Call the encoded amplitudes of the x and y channels N. Call the level of the encoded w channel W. If you then pan to (0, 0, 1), the encoded level of the z channel should also be N, but is roughly 3N, which is incorrect. It also explains what's going on perceptually: when panning around the x-y plane, it's hard to pinpoint the sound's location, which appears somewhat omnidirectional. but panning in the Z direction is more obvious, presumably because the input level is higher relative to W, so it's more directional.

2) Also, the z-axis seems to be flipped. If we send x,y,z coordinates (0, 0, 1), we hear it in the floor, and we hear (0, 0, -1) in the ceiling.

In trying to figure out why this could be happening, we realized it's possible that there are some alterations you've done to the encoder for the particular envelop systems, but we don't understand why those alterations wouldn't be happening solely inside the decoder - it would be a portability problem if alterations were in the encoder.

Happy to send more logs or debug this with you online if you like. Let me know!

- Jason
schazers@gmail.com

**Comment by ramagottfried** [comment]
Hi Jason and Natasha! 

Natasha: very sorry to miss you while you're in the bay area!! we are moving to Berlin next week so things a bit hectic!  hope to see you on the other side of the pond soon!

I have very little experience troubleshooting decoder issues -- and my understanding is that eventually there will be a Ambisonic Toolkit based decoder that will replace the current one in the Envelop server. At the moment we are using the HOALib 3D encoder and decoder, which is open source, so if you are more informed than I about the math there, you can see if you can find the problem. Otherwise, you could also email the author about it.  We are just using these objects as is, no alterations.

I vaguely remember once when I was at CCRMA testing an HOA piece of mine made with Spat, that there were also some oddities with the CCRMA decoder working with the Spat encoding. I think Fernando was maybe able to figure it out... I remember that we were able to get it working, but I totally forget what the issue was, this was a few years ago now.  Have you tested the HoaLib decoder to compare? And Spat?

I was just testing the Envelop system onsite last night, and it seems to agree with itself, but it's entirely possible that it has some eccentricities. If it's not playing nice with other systems, it's probably going to be necessary to migrate the system to a new library at some point.

best of luck, and let us know if you're able to figure it out!
Rama

**Comment by fernandoll** [comment]
Hi Rama, 
I can confirm this issue as I helped a student test Envelop this afternoon. 

This is definitely NOT a decoder issue. This is an Envelop (or library) encoder issue. 

I'm looking directly at the Ambisonics output of the Envelop encoder, bypassing any internal Envelop decoder and our decoder, examining in particular the first four harmonics (W, Y, Z, X in ACN order, or ACN components 0, 1, 2 and 3) - this is just because there are obvious problems in just the first order component encoding, I have no idea if the higher order components are correct at this point. 

I can verify by generating signals at 0 degrees azimuth and 0 degrees elevation, then at 90 degrees azimuth and 0 degrees elevation and finally at 90 degrees elevation that the order of the generated components appears to be ACN (ie: W Y Z X). I get non-zero signals in three first order components X Y and Z respectively. So the component order is good (in particular we confirm which one is Z). 

This is what I see in terms of signal level, ignoring phase for now (please ignore a possible 90 degree rotation of the soundfield around the Z axis, not all software agrees on the same coordinate system), all this is with the same "radius" or distance setting so there should be no difference in level on all three principal directions:

0 azim 0 elev:  W Y=1 Z=0 X=0
90 azim 0 elev: W Y=0 Z=0 X=1
90 elev: W Y=0 Z>1 X=0

So, for horizontal only signals in both main axis I see an (arbitrary) level of 1.0 for both X and Y, but for a vertical only signal I see a much higher level in the Z component (and 0 in both X and Y). Same as reported by Natasha above. 

In any of the existing component weighting conventions the weights of all three first order components is the same. In particular in SN3D, which Envelop implements, the weight for all four components including order 0 (W, Y, Z, X) is 1.0. 

So, signals sent along the three main axis (X/Y/Z) should generate encoded signals of equal level in all three first order components (and of course the same level on the W component for all three principal directions). 

That is not what we observe in the outputs of the Envelop encoder. The outputs of X and Y match each other approximately, but Z is clearly higher and closer in level to the W signal. 

This observation is corroborated by listening tests using our own decoder in the Listening Room set to ACN/SN3D ordering and weights as suggested by the Envelop documentation. The imaging in the horizontal plane is very vague and blurry, which corresponds to the lower level observed in X and Y with respect to Z (that is, there is a preponderance of the 0th harmonic, W,  in the horizontal plane, so all sounds are very omnidirectional). For signals coming from +90/-90 elevation the image is much more focused as would be expected from a third order system (ie: sounds as expected with one caveat below). 

One further caveat is that the Z axis is flipped (up is down and down is up). We have not experienced this issue with any other software package that outputs Ambisonics, which would suggest a phase problem with the handling of the Z component somewhere in your system. 

I'm reasonably certain that this is not an issue with our Listening Room decoder, or an error on how all signals are connected, the problem is easily observed in the output of the encoder just by looking at the relative level of the encoded components, and our decoder has been used for quite a while with encoders from different tools (AMB-plugins in Linux - FUMA, AmbiX plugins - ACN/SN3D, home brew encoders in CLM using ACN/SN3D formulas derived from ADT2's formula generator which I trust a lot, same for SuperCollider, and others). We have not experienced major problems save for the occasional 90 degree rotation around the Z axis caused by disagreements on the coordinate system (and easily fixed). 

Basically the output of the encoder, at least up to first order, does not have correct weighting and that reflects in very poor performance in the horizontal plane when using an external decoder. It does certainly not seem to be SN3D. At the very least the documentation should be updated to include this gotcha so that others trying to use the system in the same way do not waste time trying to find out why things don't work as expected. 

I don't know if the issue is in the external libraries you use or how they are used in Envelop. As I am not familiar with Max/MSP it would be helpful to get some guidance as to where to look for clues in the source tree. 

Thanks, 
-- Fernando

**Comment by ramagottfried** [comment]
Thank you @fernandoll ! this is very helpful.  ( p.s. hello from Berlin! just arrived yesterday )

Internally, Envelop currently uses the [HoaLibrary](http://www.mshparisnord.fr/hoalibrary/en/), the source code is available [here](https://github.com/CICM/HoaLibrary-Max/tree/master/Package/HoaLibrary).

Envelop is essentially a wrapper/coordination system providing authoring and performance tools for working with HOA in Ableton. I selected the HoaLib since it was written by a former graduate student at Ircam, and generally I use Spat, so it seemed like the closest open source option, but your observations are making me think we may need to reconsider. For basic usage, it seems like the HoaLib's encoder/decoder are working with each other, but I think you're right that whatever encoder/decoder we use should be standard enough to be able to integrate with other HOA systems, and that whatever format is uses should be clearly defined and corrected if it is causing issues. So, eventually we may need to consider moving to a different library. Or we could ask the author of the library to fix these issues if it makes sense to go that route. 

I'm mainly a "user" rather than "designer" of encoder/decoders, so if you can help isolate the issue maybe we can send them to the author.  Or maybe your notes above are enough already to send to them?

For more details, the encoding object Envelop currently uses is hoa.3d.map~: [here's the source](https://github.com/CICM/HoaLibrary-Max/blob/master/Max3D/hoa.3d.map_tilde.cpp)
I believe the source for encoder process function being used is [here](https://github.com/CICM/HoaLibrary-Light/blob/master/Sources/Encoder.hpp#L948)

Does anything jump out at you as an issue? 

For isolation, it would be easier to test the hoa.3d.map~ object in Max directly rather than testing with the Envelop framework.

Thanks for your help, and all the best from Berlin,
Rama

**Comment by fernandoll** [comment]
Hi Rama, 
We did more tests this afternoon (with Anna, and Wisam who knows a bit more Max/MSP) and we confirmed that this is an issue with the HOA library. We ended up testing the hoa.3d.encode (of which the map object is a wrapper) directly - even from a direct download and install - and we found the same issue (I had not read your message at that point). I have to ask about this to someone who really knows the stuff, but I don't see how that can be SN3D (which is what the source says it is). You can see the 1st order component weights being unequal in the help page for the exchanger object. The normalization transforms in Exchanger.hpp all treat XYZ the same as expected so this (X and Y different from Z) is happening somewhere else in the code. 
-- Fernando

**Comment by elsueno** [comment]
hi everyone,

does the "SN3D_Patch" downloadable from http://www.mshparisnord.fr/hoalibrary/en/downloads/max/ fix this issue?

**Comment by auralab** [comment]
I would definitely like to know if this has been fixed with the SN3D_Patch.
Did anyone ever communicate this issue to Pierre (HOALibrary author) ? It's
a critical problem with the encoding in E4L that absolutely should be
fixed.

-Andrew

**Comment by ajheller** [comment]
Both Mathias Kronlachner and contacted Pierre back in May. It is being tracked here https://github.com/CICM/HoaLibrary-Light/issues/3

I'm not a Max/MSP user, but would be glad to help anyone test the SN3D patch.

**Comment by mcslee** [comment]
Just noting here that the latest version of our project has moved over to using the ICST ambiencode~/ambidecode~ libraries instead of the HoaLibrary tools. Accordingly, will close out this issue.
